### DISCUSSION TITLE
User config to force use of defaults on specific options in Interactive mode and parameter config to disable or override interactive by default

### DISCUSSION THREAD

**Discussion by brettz9** [sapegin/mrm#332] [Ideas]
Sorry for the "spam", but too excited to see such a package as this! My last request for now, I think...

For interactive mode, how about a flag (conventionally `-y`) which would force use of the defaults supplied by CLI or config files, such that one wouldn't even have to confirm them?

**Comment by sapegin** [comment]
So, you want to see only questions without defaults?

**Comment by brettz9** [comment]
> So, you want to see only questions without defaults?

Right. Or allow config (such as a user sets-and-forgets in their config file) which indicates which tasks with defaults should (or should not) be skipped from interaction.

This becomes more desirable as the options proliferate (and I think they should be expanded so that the default tasks are robust and accommodate various styles without forcing users to make their own tasks for such basic things). For example, if we add an option for an empty `contributors: []` field to `package.json`, some projects might like to have this (I do, as besides satisfying the package.json linter, it invites future contributors to add themselves) while others who are more minimalistic, might not want this added until they need it. But even while people may have strong opinions about such small details, it may nevertheless be considered a rather trivial detail which interactive users might not wish to be bothered about upon each install if made configurable (including even for those who care about the choice but who aren't likely to vary this across their own projects).

FWIW, in working toward a tentative PR, I've run into one challenge, which is that tasks might need to check recursively for defaults, e.g., if basing a default on another option which itself has a default, I think it'd be ideal that the calculated default could get at the other option's default, so both could be auto-set without interaction. Have to explore further, but I think it should be doable.

**Comment by brettz9** [comment]
An alternative (or complementary) approach might be to allow the tasks to mark options which are more likely to be low priority as far as it not being so consequential that the average user will need to change it or be presented with the choice interactively unless they wish.

**Comment by brettz9** [comment]
So here's a possible API to allow in user config:

```js
{
  mrmSettings: {
    interactive: {
      // Enable options which a task flags as normally noninteractive so as to avoid bothering
      //   user with prompts for mundane details, e.g., adding an empty `contributors` to
      //   `package.json`; adding the option here, lets the user get a prompt without needing to
      //   directly modify a task
      enableAdvanced: ['normallyDisabledOptionToEnable', 'anotherDefaultDisabledOptionToEnable'],

      // If a project knows a default is always good for them, just use it without being bothered
      //   by a prompt and without needing to directly modify a task
      forceDefaults: ['optionToDisableFromPrompts', 'anotherOptionToOnlyUseItsDefaults'],

      // Allow the user's config file (or CLI) to override task defaults without needing to
      //   directly modify a task:
      overrideDefaults: {
        optionWithAPrexistingDefaultWeOverride: 'custom default value to be used in interactive mode',
        optionWithoutAPreexistingDefaultButForWhichWeAddOne: 'custom default value to be used in interactive mode'
      }
    }
  }
}`
```

A big advantage of this is that one doesn't have to get into tampering with task source code or additional files. Just control the values, and the values one wishes to default to (or to force defaulting, or force interaction to).

The reason for putting this proposed `interactive` idea inside of `mrmSettings`, by the way, is to have a place for objects in user config (like `type: config`) which tweak mrm behavior rather than task behavior (i.e., like the `interactive` flag but allowing for objects).

The only change to task parameters would be:

```js
module.exports.parameters = {
  contributors: {
    type: 'input',
    // ...
    noninteractiveByDefault: true
  }
};
```
